src/lib.rs

```diff
@@ -101,6 +101,128 @@ pub struct ArgParser {
+                                    , &word_lengths
+                                    , terminal_width);
+
+        let longest_words: Vec<usize> = 
+            split_into_columns( &word_lengths
+                              , columns_amt
+                              , (words.len() / columns_amt) + 1
+                              ).iter().map(longest_word).collect();
+
+        let mut words_with_space: Vec<String> = Vec::new();
+        let lines_amt = (words.len() / columns_amt) + 1;
+        let mut longest_words_rep = Vec::new();
+        for longest_word in longest_words {
+            for _ in 0..lines_amt {
+                longest_words_rep.push(longest_word.clone());
+            }
+        }
+
+        for i in 0..words.len() {
+            
+            let whitespace = " ".repeat(longest_words_rep[i] - words[i].len());
+            words[i].push_str(whitespace.as_str());
+            words_with_space.push(words[i].clone());
+        }
+
+        split_into_columns::<String>(&words_with_space, columns_amt, (words.len() / columns_amt) + 1)
+    }
+
+    /// Binary search to find the perfect amounth of columns efficiently
+    fn bin_search(min: usize, max: usize, words: &Vec<usize>, terminal_width: usize) -> usize {
+        let diff = min as isize - max as isize;
+        let fits = try_rows(words, min + (max - min) / 2, terminal_width);
+        if  diff == -1 || diff == 0 || diff == 1{
+            return min;
+        } else if fits {
+            return bin_search(min + (max - min) / 2, max, words, terminal_width);
+        } else {
+            return bin_search(min, min + (max - min) / 2, words, terminal_width);
+        }
+    }
+
+    /// Find the length of the longest word in a list
+    fn longest_word(words: &Vec<usize>) -> usize {
+        words.iter().fold(0, |x, y| {max(x, *y)})
+    }
+
+    /// splits a vector of cloneables into a vector of vectors of cloneables where lines_amt 
+    /// is the length of the outer vector and columns the max len of the inner vector
+    fn split_into_columns<T: Clone>(words: &Vec<T>, columns: usize, lines_amt: usize) -> Vec<Vec<T>> {
+        assert!(words.len() <= columns * lines_amt);
+        let mut outputs: Vec<Vec<T>> = Vec::new();
+        for _ in 0..columns {
+            outputs.push(Vec::new())
+        }
+        let mut i = 0;
+        'outer: for output in &mut outputs {
+            let mut j = 0;
+            while j < lines_amt {
+                if i >= words.len() {
+                    break 'outer;
+                }
+                output.push(words.get(i).unwrap().clone());
+                i += 1;
+                j += 1;
+            }
+        }
+        outputs
+    }
+
+    /// Tests wheter a vector of words, split into a certain number of columns fits into given width
+    fn try_rows(input: &Vec<usize>, columns: usize, width: usize) -> bool {
+
+        let lines_amt = (input.len() / columns) + 1;
```

Comment:
does it happen when the filename is longer than the terminal width?
I guess in this case columns is zero. Right now i don't have access to my coding computer. i hope i will get to fix it before friday because then i will be travelling for two weeks. maybe someone else can fix it. probable a simple `if columns == 0 {columns = 1}` will be enough.